common/map_coordinate.h

```diff
@@ -0,0 +1,231 @@
+    const double sinPhi = sin(latLon.lat);
+    const double cosPhi = cos(latLon.lat);
+    const double sinLambda = sin(latLon.lon);
+    const double cosLambda = cos(latLon.lon);
+
+    const double eSq = 2.0 * Datum::ellipsoid.f - Datum::ellipsoid.f * Datum::ellipsoid.f;  // 1st eccentricity squared ≡ (a²-b²)/a²
+    const meter_t nu = Datum::ellipsoid.a / std::sqrt(1.0 - eSq * sinPhi * sinPhi);         // radius of curvature in prime vertical
+
+    return Cartesian<Datum>{(nu + h) * cosPhi * cosLambda,
+                            (nu + h) * cosPhi * sinLambda,
+                            (nu * (1.0 - eSq) + h) * sinPhi};
+}
+
+// Transform cartesian coordinates from WGS84 into desired space
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
+template<typename Datum>
+Cartesian<Datum> transformCartesian(const Cartesian<WGS84> &c)
+{
+    const double s1 = Datum::transform.s / 1E6 + 1.0;            // scale: normalise parts-per-million to (s+1)
+
+    const double rx = Datum::transform.rx.value();
+    const double ry = Datum::transform.ry.value();
+    const double rz = Datum::transform.rz.value();
+
+    // apply transform
+    return Cartesian<Datum>{Datum::transform.tx + c.x * s1 - c.y * rz + c.z * ry,
+                            Datum::transform.ty + c.x * rz + c.y * s1 - c.z * rx,
+                            Datum::transform.tz - c.x * ry + c.y * rx + c.z * s1};
+}
+
+// Convert cartesian coordinates back to latitude and longitude
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
+template<typename Datum>
+LatLon<Datum> cartesianToLatLon(const Cartesian<Datum> &c)
+{
+    using namespace units::angle;
+    using namespace units::length;
+    using namespace units::math;
+    using namespace units::literals;
+
+    const double e2 = 2.0 * Datum::ellipsoid.f - Datum::ellipsoid.f * Datum::ellipsoid.f;    // 1st eccentricity squared ≡ (a²-b²)/a²
+    const double epsilon2 = e2 / (1.0 - e2);                                                // 2nd eccentricity squared ≡ (a²-b²)/b²
+    const meter_t p = sqrt(c.x * c.x + c.y * c.y);  // distance from minor axis
+    const meter_t r = sqrt(p * p + c.z * c.z);      // polar radius
+
+    // parametric latitude (Bowring eqn 17, replacing tanβ = z·a / p·b)
+    const double tanBeta = (Datum::ellipsoid.b * c.z) / (Datum::ellipsoid.a * p) * (1.0 + epsilon2 * Datum::ellipsoid.b / r);
+    const double sinBeta = tanBeta / std::sqrt(1.0 + tanBeta * tanBeta);
+    const double cosBeta = sinBeta / tanBeta;
+
+    // geodetic latitude (Bowring eqn 18: tanφ = z+ε²bsin³β / p−e²cos³β)
+    const degree_t phi = std::isnan(cosBeta) ? 0.0_rad : atan2(c.z + epsilon2 * Datum::ellipsoid.b * sinBeta * sinBeta * sinBeta,
+                                                               p - e2 * Datum::ellipsoid.a * cosBeta * cosBeta * cosBeta);
```

Comment:
Am I right in thinking ``std::isnan(cosBeta)`` is the same as ``tanBeta <= 0`` here? If so, that might be a bit clearer, rather than having to try to remember which floating point operations can result in NaN.